cmd/feature.go

```diff
@@ -0,0 +1,194 @@
+	Long: `
+cozy-stack feature sets displays the feature sets coming from the manager.
+
+It can also take a list of sets that will replace the previous list (no merge).
+
+All the sets can be removed by setting an empty list ('').
+`,
+	Example: `$ cozy-stack feature sets --domain cozy.tools:8080 'set1 set2'`,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		if flagDomain == "" {
+			errPrintfln("%s", errMissingDomain)
+			return cmd.Usage()
+		}
+		c := newAdminClient()
+		req := request.Options{
+			Method: "GET",
+			Path:   fmt.Sprintf("/instances/%s/feature/sets", flagDomain),
+		}
+		if len(args) > 0 {
+			list := args
+			if len(args) == 1 {
+				list = strings.Fields(args[0])
+			}
+			buf, err := json.Marshal(list)
+			if err != nil {
+				return err
+			}
+			req.Method = "PUT"
+			req.Body = bytes.NewReader(buf)
+		}
+		res, err := c.Req(&req)
+		if err != nil {
+			return err
+		}
+		defer res.Body.Close()
+		var sets []string
+		if err = json.NewDecoder(res.Body).Decode(&sets); err != nil {
+			return err
+		}
+		for _, set := range sets {
+			fmt.Printf("- %v\n", set)
+		}
+		return nil
+	},
+}
+
+var featureContextCmd = &cobra.Command{
+	Use:   "context <context-name>",
+	Short: `Display and update the feature flags for a context`,
+	Long: `
+cozy-stack feature defaults displays the feature flags for a context.
+
+It can also create, update, or remove flags (with a ratio and value).
+
+To remove a flag, set it to an empty array (or null).
+`,
+	Example: `$ cozy-stack feature context beta '{"set_this_flag": [{"ratio": 0.1, "value": 1}, {"ratio": 0.9, "value": 2}] }'`,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		if len(args) == 0 || args[0] == "" {
+			return cmd.Usage()
+		}
+		c := newAdminClient()
+		req := request.Options{
+			Method: "GET",
+			Path:   fmt.Sprintf("/instances/feature/contexts/%s", args[0]),
+		}
+		if len(args) > 1 {
+			req.Method = "PATCH"
+			req.Body = strings.NewReader(args[1])
+		}
+		res, err := c.Req(&req)
+		if err != nil {
+			return err
+		}
+		defer res.Body.Close()
+		var obj map[string]json.RawMessage
+		if err = json.NewDecoder(res.Body).Decode(&obj); err != nil {
+			return err
+		}
+		for k, v := range obj {
+			fmt.Printf("- %s: %s\n", k, string(v))
+		}
+		return nil
+	},
+}
+
+var featureDefaultCmd = &cobra.Command{
+	Use:   "defaults",
+	Short: `Display and update the default values for feature flags`,
+	Long: `
+cozy-stack feature defaults displays the default values for feature flags.
+
+It can also take a list of flags to update.
+
+If you give a null value, the flag will be removed.
+`,
+	Example: `$ cozy-stack feature defaults '{"add_this_flag": true, "remove_this_flag": null}'`,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		c := newAdminClient()
+		req := request.Options{
+			Method: "GET",
+			Path:   "/instances/feature/defaults",
+		}
+		if len(args) > 0 {
+			req.Method = "PATCH"
+			req.Body = strings.NewReader(args[0])
```

Comment:
IMHO inputs should be validated. maybe not here but at least in the `/instances/feature/defaults` route

Reply:
What validation do you expect? Do you have an example of invalid input that we should reject (and is not yet rejected)?

Reply:
req.Body is supposed to be valid json.
`foo{baf"baz}` is no valid json.
What do we get when monkey-typing parameters? is that a clean error or a global crash?

Reply:
We already check that:

```
$ cozy-stack feature defaults '{baf"baz}' 
Error: Unqualified error: invalid character 'b' looking for beginning of object key string
exit status 1
```

Reply:
oh, ok, I didn't see that, so forget about it.